flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -0,0 +1,1000 @@
+ *
+ * <p>Implementation note: This producer is a hybrid between a regular regular
+ * {@link org.apache.flink.streaming.api.functions.sink.SinkFunction} (a) and a custom operator (b).
+ *
+ * <p>Details about approach (a):
+ *  Because of regular {@link org.apache.flink.streaming.api.functions.sink.SinkFunction} APIs limitations, this
+ *  variant do not allow accessing the timestamp attached to the record.
+ *
+ * <p>Details about approach (b):
+ *  Kafka 0.11 supports writing the timestamp attached to a record to Kafka. When using the
+ *  {@link FlinkKafkaProducer011#writeToKafkaWithTimestamps} method, the Kafka producer can access the internal
+ *  record timestamp of the record and write it to Kafka.
+ *
+ * <p>All methods and constructors in this class are marked with the approach they are needed for.
+ */
+public class FlinkKafkaProducer011<IN>
+		extends TwoPhaseCommitSinkFunction<IN, FlinkKafkaProducer011.KafkaTransactionState> {
+
+	/**
+	 *  Semantics that can be chosen.
+	 *  <li>{@link #EXACTLY_ONCE}</li>
+	 *  <li>{@link #AT_LEAST_ONCE}</li>
+	 *  <li>{@link #NONE}</li>
+	 */
+	public enum Semantic {
+		/**
+		 * Semantic.EXACTLY_ONCE the Flink producer will write all messages in a Kafka transaction that will be
+		 * committed to the Kafka on a checkpoint.
+		 *
+		 * <p>In this mode {@link FlinkKafkaProducer011} sets up a pool of {@link FlinkKafkaProducer}. Between each
+		 * checkpoint there is created new Kafka transaction, which is being committed on
+		 * {@link FlinkKafkaProducer011#notifyCheckpointComplete(long)}. If checkpoint complete notifications are
+		 * running late, {@link FlinkKafkaProducer011} can run out of {@link FlinkKafkaProducer}s in the pool. In that
+		 * case any subsequent {@link FlinkKafkaProducer011#snapshotState(FunctionSnapshotContext)} requests will fail
+		 * and {@link FlinkKafkaProducer011} will keep using the {@link FlinkKafkaProducer} from previous checkpoint.
+		 * To decrease chances of failing checkpoints there are three options:
+		 * <li>decrease number of max concurrent checkpoints</li>
+		 * <li>make checkpoints more reliable (so that they complete faster)</li>
+		 * <li>increase delay between checkpoints</li>
+		 * <li>increase size of {@link FlinkKafkaProducer}s pool</li>
+		 */
+		EXACTLY_ONCE,
+		/**
+		 * Semantic.AT_LEAST_ONCE the Flink producer will wait for all outstanding messages in the Kafka buffers
+		 * to be acknowledged by the Kafka producer on a checkpoint.
+		 */
+		AT_LEAST_ONCE,
+		/**
+		 * Semantic.NONE means that nothing will be guaranteed. Messages can be lost and/or duplicated in case
+		 * of failure.
+		 */
+		NONE
+	}
+
+	private static final Logger LOG = LoggerFactory.getLogger(FlinkKafkaProducerBase.class);
+
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * Default number of KafkaProducers in the pool. See {@link Semantic#EXACTLY_ONCE}.
+	 */
+	public static final int DEFAULT_KAFKA_PRODUCERS_POOL_SIZE = 5;
```

Comment:
As I remember the reason was that it is not easy/not possible at the moment to get this information in the operator. It should be a follow up work. Regardless of this, code of this operator would look the same (because we don't have guarantees for the `notifyCheckpointComplete` to always reach us on time).